`enable-iam-auth` property is ineffective for R2DBC

Using Spring Cloud GCP 3.4.2, connecting to an IAM-enabled Google Cloud SQL instance over R2DBC does not work out of the box.

Using the following `application.yml`:

```yaml
spring:
  cloud:
    gcp:
      project-id: pro-tetris-f2f
      sql:
        database-name: tetris
        instance-connection-name: pro-tetris-f2f:europe-west4:tetris-csql-i
        enable-iam-auth: true
  r2dbc:
    username: tetris-sa@pro-tetris-f2f.iam
```

results in the following failure:

```
Error: SEVERITY_LOCALIZED=FATAL, SEVERITY_NON_LOCALIZED=FATAL, CODE=28000, MESSAGE=Cloud SQL IAM service account authentication failed for user "tetris-sa@pro-tetris-f2f.iam", FILE=auth.c, LINE=345, ROUTINE=auth_failed
```

It works as expected once `ENABLE_IAM_AUTH=true` is added to the R2DBC URL by means of adding

1. either `spring.r2dbc.properties.ENABLE_IAM_AUTH: true` property
2. or `spring.r2dbc.url: r2dbc:gcp:postgres://pro-tetris-f2f:europe-west4:tetris-csql-i/tetris` property

We expected this flag to be automatically set due to `spring.cloud.gcp.sql.enable-iam-auth: true` property.

@diegomarquezp, I will share the project files for a minimal reproduction example below.

`pom.xml`:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">

  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-parent</artifactId>
    <version>3.0.2</version>
  </parent>

  <groupId>com.bol.cloud.google</groupId>
  <artifactId>spring-cloud-gcp-issue-1513-demo</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <name>spring-cloud-gcp-issue-1513-demo</name>
  <description>Spring Cloud GCP issue #1513 reproduction</description>

  <properties>
    <java.version>17</java.version>
  </properties>

  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>com.google.cloud</groupId>
        <artifactId>spring-cloud-gcp-dependencies</artifactId>
        <version>3.4.2</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>

  <dependencies>
    <dependency>
      <groupId>com.google.cloud</groupId>
      <artifactId>spring-cloud-gcp-starter-sql-postgres-r2dbc</artifactId>
    </dependency>
  </dependencies>

</project>

```

`SpringCloudGcpIssue1513DemoApplication.java`:

```
package com.bol.cloud.google.springcloudgcpissue1513demo;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class SpringCloudGcpIssue1513DemoApplication {

	public static void main(String[] args) {
		SpringApplication.run(SpringCloudGcpIssue1513DemoApplication.class, args);
	}

}
```

`ConnectionCheck.java`:

```
package com.bol.cloud.google.springcloudgcpissue1513demo;

import io.r2dbc.spi.ConnectionFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.event.ApplicationStartedEvent;
import org.springframework.context.ApplicationListener;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

import java.time.Duration;

@Component
public class ConnectionCheck implements ApplicationListener<ApplicationStartedEvent> {

    @Autowired
    private ConnectionFactory connectionFactory;

    @Override
    public void onApplicationEvent(ApplicationStartedEvent ignored) {
        Mono
                .from(connectionFactory.create())
                .flatMapMany(connection -> connection.createStatement("SELECT 1").execute())
                .flatMap(result -> result.map((row, meta) -> row.get(0, Integer.class)))
                .doOnNext(updatedRowCount -> System.out.format("row: %d%n", updatedRowCount))
                .then()
                .block(Duration.ofSeconds(5));
    }

}
```

`application.yml`:

```
spring:
  r2dbc:
    username: "vyazici@bol.com"
    properties:
      ENABLE_IAM_AUTH: true
  cloud:
    gcp:
      project-id: "bolcom-stg-axlexamplecloud-d2c"
      sql:
        database-name: "examplecloud"
        instance-connection-name: "bolcom-stg-axlexamplecloud-d2c:europe-west4:examplecloud-csql-i"
        enable-iam-auth: true
```

The project successfully queries the Cloud SQL instance over IAM using the above `application.yml` configuration. But...

When I remove `r2dbc.properties.ENABLE_IAM_AUTH: true` property, I get the following error:

```
Caused by: java.lang.IllegalArgumentException: Password must not be null
	at io.r2dbc.postgresql.util.Assert.requireNonNull(Assert.java:71) ~[r2dbc-postgresql-0.8.13.RELEASE.jar:0.8.13.RELEASE]
	at io.r2dbc.postgresql.PostgresqlConnectionFactory.getAuthenticationHandler(PostgresqlConnectionFactory.java:242) ~[r2dbc-postgresql-0.8.13.RELEASE.jar:0.8.13.RELEASE]
	at io.r2dbc.postgresql.client.StartupMessageFlow.lambda$exchange$0(StartupMessageFlow.java:76) ~[r2dbc-postgresql-0.8.13.RELEASE.jar:0.8.13.RELEASE]
	at reactor.core.publisher.FluxHandle$HandleConditionalSubscriber.onNext(FluxHandle.java:326) ~[reactor-core-3.5.2.jar:3.5.2]
	... 43 common frames omitted
```

Then I add `spring.r2dbc.password: RequiredButNotUsed` property, this time I get an authentication error:

```
io.r2dbc.postgresql.ExceptionFactory$PostgresqlAuthenticationFailure: Cloud SQL IAM user authentication failed for user "vyazici@bol.com"
	at io.r2dbc.postgresql.ExceptionFactory.createException(ExceptionFactory.java:75) ~[r2dbc-postgresql-0.8.13.RELEASE.jar:0.8.13.RELEASE]
        ...
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:139) ~[reactor-core-3.5.2.jar:3.5.2]
		at reactor.core.publisher.Mono.block(Mono.java:1734) ~[reactor-core-3.5.2.jar:3.5.2]
		at com.bol.cloud.google.springcloudgcpissue1513demo.ConnectionCheck.onApplicationEvent(ConnectionCheck.java:26) ~[classes/:na]
		at com.bol.cloud.google.springcloudgcpissue1513demo.ConnectionCheck.onApplicationEvent(ConnectionCheck.java:12) ~[classes/:na]
               ...
```

Yes the error message is different than what I reported initially. It has changed after appending `--update-adc` to `gcloud auth login`. Nevertheless, my point still stands: without adding either `spring.r2dbc.properties.ENABLE_IAM_AUTH: true` or `spring.r2dbc.url`, IAM-based authentication doesn't work.

--- a/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/R2dbcCloudSqlEnvironmentPostProcessor.java
+++ b/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/R2dbcCloudSqlEnvironmentPostProcessor.java
@@ -67,6 +67,14 @@ public class R2dbcCloudSqlEnvironmentPostProcessor implements EnvironmentPostPro
 
       CredentialsPropertiesSetter.setCredentials(sqlProperties,
           propertiesRetriever.getGcpProperties());
+
+      if (sqlProperties.isEnableIamAuth()) {
+        environment
+            .getPropertySources()
+            .addFirst(
+                new MapPropertySource("CLOUD_SQL_R2DBC_ENABLE_IAM_AUTH",
+                    Map.of("spring.r2dbc.properties", Map.of("ENABLE_IAM_AUTH", "true"))));
+      }
     }
   }
 

--- a/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/sql/R2dbcCloudSqlEnvironmentPostProcessorTests.java
+++ b/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/sql/R2dbcCloudSqlEnvironmentPostProcessorTests.java
@@ -52,45 +52,32 @@ class R2dbcCloudSqlEnvironmentPostProcessorTests {
 
   @Test
   void testSetR2dbcProperty_postgres() {
-    verifyThatCorrectUrlAndUsernameSet(
-        new String[] {"io.r2dbc.postgresql"},
-        "postgres",
-        "r2dbc:gcp:postgres://my-project:region:my-instance/my-database");
+    this.contextRunner
+        .withPropertyValues(
+            "spring.cloud.gcp.sql.databaseName=my-database",
+            "spring.cloud.gcp.sql.instanceConnectionName=my-project:region:my-instance")
+        .withClassLoader(excludeDriverPackages(new String[] {"io.r2dbc.postgresql"}))
+        .run(
+            context -> {
+              assertThat(context.getEnvironment().getProperty("spring.r2dbc.url"))
+                  .isEqualTo("r2dbc:gcp:postgres://my-project:region:my-instance/my-database");
+              assertThat(context.getEnvironment().getProperty("spring.r2dbc.username"))
+                  .isEqualTo("postgres");
+            });
   }
 
-  /**
-   * Verifies that correct database properties got injected into context, given a passed-in list of
-   * packages to retain on the classpath.
-   *
-   * @param driverPackagesToInclude a list of driver packages to keep on the classpath
-   * @param username expected {@code spring.r2dbc.username} value to verify
-   * @param url expected {@code spring.r2dbc.username} value to verify
-   */
+  @Test
   @SuppressWarnings("unchecked")
-  private void verifyThatCorrectUrlAndUsernameSet(
-      String[] driverPackagesToInclude,
-      String username,
-      String url) {
-    // Because `FilteredClassLoader` accepts a list of packages to remove from classpath,
-    // `driverPackagesToInclude` is used to calculate the inverse list of packages to _exclude_.
-    Set<String> driverPackagesToExclude = new HashSet<>(List.of("io.r2dbc.postgresql"));
-    Arrays.asList(driverPackagesToInclude).forEach(driverPackagesToExclude::remove);
-
+  void testEnableIamAuthPresent() {
     this.contextRunner
         .withPropertyValues(
             "spring.cloud.gcp.sql.databaseName=my-database",
             "spring.cloud.gcp.sql.instanceConnectionName=my-project:region:my-instance",
             "spring.cloud.gcp.sql.enable-iam-auth=true")
-        .withClassLoader(new FilteredClassLoader(driverPackagesToExclude.toArray(new String[0])))
+        .withClassLoader(excludeDriverPackages(new String[] {"io.r2dbc.postgresql"}))
         .run(
-            context -> {
-              assertThat(context.getEnvironment().getProperty("spring.r2dbc.url"))
-                  .isEqualTo(url);
-              assertThat(context.getEnvironment().getProperty("spring.r2dbc.username"))
-                  .isEqualTo(username);
-              assertThat(context.getEnvironment().getProperty("spring.r2dbc.properties", Map.class, Map.of()))
-                  .isEqualTo(Map.of("ENABLE_IAM_AUTH", "true"));
-            });
+            context -> assertThat(context.getEnvironment().getProperty("spring.r2dbc.properties", Map.class, Map.of()))
+                .isEqualTo(Map.of("ENABLE_IAM_AUTH", "true")));
   }
 
   @Test
@@ -121,5 +108,22 @@ class R2dbcCloudSqlEnvironmentPostProcessorTests {
         .run(
             context -> assertThat(r2dbcPostProcessor.getEnabledDatabaseType(context.getEnvironment()))
                 .isNull());
+  }
+
+  /**
+   * Returns a class loader which excludes certain driver packages from the classpath
+   * unless they are in {@code driverPackagesToInclude}.
+   *
+   * @param driverPackagesToInclude A list of driver packages to keep on the classpath
+   * @return A class loader with a set of driver packages that are excluded from the classpath
+   */
+  private FilteredClassLoader excludeDriverPackages(String[] driverPackagesToInclude) {
+    // Because `FilteredClassLoader` accepts a list of packages to remove from classpath,
+    // `driverPackagesToInclude` is used to calculate the inverse list of packages to _exclude_.
+    Set<String> driverPackagesToExclude = new HashSet<>(List.of("io.r2dbc.postgresql"));
+    Arrays.stream(driverPackagesToInclude)
+        .forEach(driverPackagesToExclude::remove);
+
+    return new FilteredClassLoader(driverPackagesToExclude.toArray(new String[0]));
   }
 }

--- a/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/sql/R2dbcCloudSqlEnvironmentPostProcessorTests.java
+++ b/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/sql/R2dbcCloudSqlEnvironmentPostProcessorTests.java
@@ -21,6 +21,7 @@ import static org.assertj.core.api.Assertions.assertThat;
 import java.util.Arrays;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 import org.junit.jupiter.api.Test;
 import org.springframework.boot.SpringApplication;
@@ -65,6 +66,7 @@ class R2dbcCloudSqlEnvironmentPostProcessorTests {
    * @param username expected {@code spring.r2dbc.username} value to verify
    * @param url expected {@code spring.r2dbc.username} value to verify
    */
+  @SuppressWarnings("unchecked")
   private void verifyThatCorrectUrlAndUsernameSet(
       String[] driverPackagesToInclude,
       String username,
@@ -77,7 +79,8 @@ class R2dbcCloudSqlEnvironmentPostProcessorTests {
     this.contextRunner
         .withPropertyValues(
             "spring.cloud.gcp.sql.databaseName=my-database",
-            "spring.cloud.gcp.sql.instanceConnectionName=my-project:region:my-instance")
+            "spring.cloud.gcp.sql.instanceConnectionName=my-project:region:my-instance",
+            "spring.cloud.gcp.sql.enable-iam-auth=true")
         .withClassLoader(new FilteredClassLoader(driverPackagesToExclude.toArray(new String[0])))
         .run(
             context -> {
@@ -85,6 +88,8 @@ class R2dbcCloudSqlEnvironmentPostProcessorTests {
                   .isEqualTo(url);
               assertThat(context.getEnvironment().getProperty("spring.r2dbc.username"))
                   .isEqualTo(username);
+              assertThat(context.getEnvironment().getProperty("spring.r2dbc.properties", Map.class, Map.of()))
+                  .isEqualTo(Map.of("ENABLE_IAM_AUTH", "true"));
             });
   }
 
@@ -96,10 +101,8 @@ class R2dbcCloudSqlEnvironmentPostProcessorTests {
                 "com.google.cloud.sql.core.GcpConnectionFactoryProviderMysql",
                 "com.google.cloud.sql.core.GcpConnectionFactoryProviderPostgres"))
         .run(
-            context -> {
-              assertThat(r2dbcPostProcessor.getEnabledDatabaseType(context.getEnvironment()))
-                  .isNull();
-            });
+            context -> assertThat(r2dbcPostProcessor.getEnabledDatabaseType(context.getEnvironment()))
+                .isNull());
   }
 
   @Test
@@ -107,10 +110,8 @@ class R2dbcCloudSqlEnvironmentPostProcessorTests {
     this.contextRunner
         .withClassLoader(new FilteredClassLoader("io.r2dbc.spi.ConnectionFactory"))
         .run(
-            context -> {
-              assertThat(r2dbcPostProcessor.getEnabledDatabaseType(context.getEnvironment()))
-                  .isNull();
-            });
+            context -> assertThat(r2dbcPostProcessor.getEnabledDatabaseType(context.getEnvironment()))
+                .isNull());
   }
 
   @Test
@@ -118,9 +119,7 @@ class R2dbcCloudSqlEnvironmentPostProcessorTests {
     this.contextRunner
         .withPropertyValues("spring.cloud.gcp.sql.r2dbc.enabled=false")
         .run(
-            context -> {
-              assertThat(r2dbcPostProcessor.getEnabledDatabaseType(context.getEnvironment()))
-                  .isNull();
-            });
+            context -> assertThat(r2dbcPostProcessor.getEnabledDatabaseType(context.getEnvironment()))
+                .isNull());
   }
 }

--- a/spring-cloud-gcp-samples/spring-cloud-gcp-sql-postgres-r2dbc-sample/src/test/java/com/example/SqlR2dbcPostgresSampleApplicationIntegrationTests.java
+++ b/spring-cloud-gcp-samples/spring-cloud-gcp-sql-postgres-r2dbc-sample/src/test/java/com/example/SqlR2dbcPostgresSampleApplicationIntegrationTests.java
@@ -35,7 +35,7 @@ import org.springframework.http.ResponseEntity;
     properties = {
       "spring.cloud.gcp.sql.databaseName=code_samples_test_r2dbc_db",
       "spring.cloud.gcp.sql.instanceConnectionName=spring-cloud-gcp-ci:us-central1:testpostgres",
-      "spring.r2dbc.password=test"
+      "spring.cloud.gcp.sql.enable-iam.auth=true"
     })
 @EnabledIfSystemProperty(named = "it.cloudsql", matches = "true")
 class SqlR2dbcPostgresSampleApplicationIntegrationTests {

